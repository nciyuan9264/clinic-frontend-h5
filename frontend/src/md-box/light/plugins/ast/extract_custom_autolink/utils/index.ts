export { HyperNodeType, getHyperLinkNodes } from './parser';

export const completeProtocol = (raw: string) => {
  if (raw.match(/https?:\/\//)) {
    return raw;
  }
  return `https://${raw}`;
};
