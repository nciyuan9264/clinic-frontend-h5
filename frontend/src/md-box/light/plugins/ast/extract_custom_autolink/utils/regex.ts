// -> refer @ea/url-recognition -> refer 飞书云文档
const PROTOCOL = 'https?|s?ftp|ftps|nfs|ssh';
const LINK_SUFFIX =
  "\\b([/#?][\\-A-Z0-9@:%_+.~#?&'/=;$,!\\*\\[\\]\\(\\){}^|<>]*[-A-Z0-9&@#/%=~\\(\\)_|]?)?";
const LOCALHOST_REG = /localhost(:[0-9]{2,5})?/;
const IP_ADDRESS_REG =
  // eslint-disable-next-line max-len
  /(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}(2[0-4][0-9]|25[0-5]|1[0-9]{2}|[1-9][0-9]|[0-9])(:[0-9]{2,5})?/;
const IP_REG = new RegExp(
  `((((${PROTOCOL})://)?(${LOCALHOST_REG.source}))|(((${PROTOCOL})://)(${IP_ADDRESS_REG.source})))${LINK_SUFFIX}`,
  'gi',
);
const URL_HOST_BODY = '([-A-Z0-9%_+~#@][\\-A-Z:0-9%_+~#@]{0,256}\\.){1,50}';
const ANY_TOP_DOMAIN = '[a-z\\-]{2,15}';
const TOP_DOMAIN =
  'com|org|net|int|edu|gov|mil|ac|ad|ae|af|ag|ai|al|am|an|ao|aq|ar|as|at|au|aw|ax|az|ba|bb|bd|be|bf|bg|bh|bi|bj|bm|bn|bo|br|bs|bt|bv|bw|by|bz|ca|cc|cf|cg|ch|ci|ck|cl|cm|cn|co|cr|cu|cv|cw|cx|cy|cz|de|dj|dk|dm|do|dz|ec|ee|eg|eh|er|es|et|eu|fi|fj|fk|fm|fo|fr|ga|gb|gd|ge|gf|gh|gi|gl|gm|gn|gp|gr|gs|gt|gu|gw|gy|hk|hm|hn|hr|ht|hu|id|ie|il|im|in|io|iq|ir|is|it|jm|jo|jp|ke|kg|kh|ki|km|kn|kp|kr|kw|ky|kz|la|lb|lc|li|lk|lr|ls|lt|lu|lv|ly|ma|mc|md|me|mf|mg|mh|mk|ml|mm|mn|mo|mp|mq|mr|ms|mt|mu|mv|mw|mx|my|mz|na|nc|ne|nf|ng|ni|nl|no|np|nr|nu|nz|om|pa|pe|pf|pg|ph|pk|pl|pm|pn|pr|ps|pt|pw|py|qa|re|ro|rs|ru|rw|sa|sb|sc|sd|se|sg|sh|si|sj|sk|sl|sm|sn|so|sr|ss|st|su|sv|sx|sy|sz|tc|td|tf|tg|th|tj|tk|tl|tm|tn|to|tp|tr|tt|tv|tw|tz|ua|ug|uk|um|us|uy|uz|va|vc|ve|vg|vi|vn|vu|wf|ws|ye|yt|za|zm|zw|site|top|wtf|xxx|xyz|cloud|engineering|help|one';
const TOP_DOMAIN_LITE = 'com|cn|tk|de|net|org|uk|info|nl|ru|hk';
// 由于顶级域名不全，故更改为带 协议头的url 忽略上面顶级域名，直接匹配
const URL_REG = new RegExp(
  // eslint-disable-next-line max-len
  `(((${PROTOCOL}):\\/\\/${URL_HOST_BODY}${ANY_TOP_DOMAIN})|(${URL_HOST_BODY}(${TOP_DOMAIN_LITE})))(:[0-9]{2,5})?${LINK_SUFFIX}`,
  'gi',
);
// mailto的贪婪匹配在遇到极长的字符串（无空格的英文）会很慢，需要限长。
const EMAIL_URL_BODY = "[\\w.!#$%&'*+-/=?^_\\`{|}~]{1,2000}@[A-Za-z0-9_.-]+\\.";
const LENGTH_LIMITED_EMAIL_REG = new RegExp(
  `^(?!data:)((mailto:${EMAIL_URL_BODY}${ANY_TOP_DOMAIN})|(${EMAIL_URL_BODY}(${TOP_DOMAIN})))\\b`,
  'gi',
);

export const urlRegExp = new RegExp(
  `(${IP_REG.source})|(${URL_REG.source})|(${LENGTH_LIMITED_EMAIL_REG.source})`,
  'gi',
);
