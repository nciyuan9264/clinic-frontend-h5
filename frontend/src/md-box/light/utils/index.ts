export * from './micromark';
export * from './marked';
export * from './renderer';
export * from './common';
export * from './hop';
export * from './plugin';
