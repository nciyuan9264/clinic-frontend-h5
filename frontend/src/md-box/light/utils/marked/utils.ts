/** 源自 markedjs 源码 */

const escapeTest = /[&<>"']/;
const escapeReplace = new RegExp(escapeTest.source, 'g');
const escapeTestNoEncode = /[<>"']|&(?!(#\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\w+);)/;
const escapeReplaceNoEncode = new RegExp(escapeTestNoEncode.source, 'g');

const escapeReplacements: Record<string, string> = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#39;',
};

export const escapeMarkdownText = (html: string, encode = false) => {
  const getEscapeReplacement = (ch: string) => escapeReplacements[ch];

  if (encode) {
    if (escapeTest.test(html)) {
      return html.replace(escapeReplace, getEscapeReplacement);
    }
  } else {
    if (escapeTestNoEncode.test(html)) {
      return html.replace(escapeReplaceNoEncode, getEscapeReplacement);
    }
  }
  return html;
};
