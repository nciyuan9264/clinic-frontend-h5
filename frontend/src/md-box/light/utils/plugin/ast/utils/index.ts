export * from './update_ast';
