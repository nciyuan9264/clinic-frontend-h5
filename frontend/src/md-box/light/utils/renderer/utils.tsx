import React, { createElement } from 'react';

import { isRegExp } from 'lodash-es';
import { DOMNode, Element, Text, domToReact } from 'html-react-parser';

import { RenderFunction } from '../../type';

export const isTextType = (node: DOMNode): node is Text => {
  return node.type === 'text';
};

export const isEndlineText = (node: DOMNode): node is Text => {
  return isTextType(node) && /^\n+$/.test(getInnerText(node));
};

/**
 * 此处断言的是Element的子类型：即为A标签的类型，但是Element不可进一步进行类型收窄，因此
 * 返回node is Element
 */
export const isElementType = (
  node: DOMNode,
  ...names: (string | RegExp)[]
): node is Element => {
  if (!('attribs' in node) || typeof node.attribs !== 'object') {
    return false;
  }
  return (
    node.type === 'tag' &&
    (!names?.length ||
      names.some((item) => {
        if (isRegExp(item)) {
          return item.test(node.name);
        } else {
          return item === node.name;
        }
      }))
  );
};

export const getInnerText = (node: DOMNode): string => {
  if (isTextType(node)) {
    return node.data;
  }
  if (isElementType(node)) {
    return node.children.map(getInnerText).join('');
  }
  return '';
};

/**
 * 检测字符或字符串是否包含中文字符
 * https://github.com/anton-bot/contains-chinese/blob/master/index.js#L7
 * @param char 字符或字符串
 * @returns 是否包含中文字符
 */
export const detectIsChinese = (char: string) => {
  const HAN_REGEX =
    // eslint-disable-next-line max-len
    /[\u2E80-\u2E99\u2E9B-\u2EF3\u2F00-\u2FD5\u3005\u3007\u3021-\u3029\u3038-\u303B\u3400-\u4DB5\u4E00-\u9FD5\uF900-\uFA6D\uFA70-\uFAD9]/;
  return HAN_REGEX.test(char);
};

export const renderDom: RenderFunction<false> = (node) => {
  return <>{domToReact([node])}</>;
};

export const renderReactElement = <P extends Record<string, unknown>>(
  tag: string,
  props: P,
  children?: React.ReactNode[],
) => {
  if (props.className === 'markdown-root') {
    return <>{children}</>;
  } else {
    return (
      <>{createElement(tag, props, children?.length ? children : undefined)}</>
    );
  }
};
