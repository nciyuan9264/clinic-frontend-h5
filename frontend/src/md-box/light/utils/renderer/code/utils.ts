import { DOMNode, Element } from 'html-react-parser';

import { isElementType } from '../utils';

const DEFAULT_LANGUAGE = 'plaintext';

const getCodeElement = (node: DOMNode): Element | undefined => {
  if (isElementType(node, 'code')) {
    return node;
  }

  if (isElementType(node, 'pre')) {
    return node.childNodes.find((childNode): childNode is Element =>
      isElementType(childNode, 'code'),
    );
  }
};

export const getLanguageFromElement = (
  node: DOMNode,
  defaultLanguage = DEFAULT_LANGUAGE,
): string => {
  const codeElement = getCodeElement(node);

  if (codeElement) {
    return (
      codeElement.attribs?.class?.match(/language-(.*)/)?.[1] || defaultLanguage
    );
  }

  return defaultLanguage;
};

export const getMetaFromElement = (node: DOMNode, removeBuiltIn = true) => {
  const codeElement = getCodeElement(node);

  const metaString = codeElement?.attribs['data-meta'];

  if (!metaString) {
    return;
  }

  if (!removeBuiltIn) {
    return metaString;
  }

  return (
    metaString
      .split(/\s+/)
      .filter((item) => !item.match(/^__[^_].*/))
      .join(' ') || undefined
  );
};

export const getIndentFromCodeElememt = (node: DOMNode) => {
  const metaString = getMetaFromElement(node, false);

  if (!metaString) {
    return 0;
  }

  const matchResult = metaString.match(
    /(\s|^)__indent=(?<indent>[0-9]+)(\s|$)/,
  );

  if (!matchResult) {
    return 0;
  }

  const indentValue = Number(matchResult.groups?.indent);

  return isNaN(indentValue) ? 0 : indentValue;
};
