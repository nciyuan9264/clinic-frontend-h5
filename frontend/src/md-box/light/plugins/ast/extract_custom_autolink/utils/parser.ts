import emailValidator from 'email-validator';

import { urlRegExp } from './regex';

export enum HyperNodeType {
  text,
  link,
}

export type HyperNode =
  | {
      type: HyperNodeType.text;
      text: string;
    }
  | {
      type: HyperNodeType.link;
      url: string;
    };

export type HyperNodeTitled = HyperNode & {
  title: string;
};

const openAndCloseCharPairs: Record<string, string> = {
  '(': ')',
  '[': ']',
  '{': '}',
  '<': '>',
};
const openChars = Object.keys(openAndCloseCharPairs);
const closeChars = Object.values(openAndCloseCharPairs);
const unreasonableSuffix = ['.', ',', '!', '?', "'", ':', ';', '@', '#'];

/**
 * 分割合理的URL
 * @returns 返回字符串中合理的URL的长度，范围[0, input.length]
 */
const getAReasonableUrl = (input: string) => {
  let index = -1;
  const stack: Array<{
    char: string;
    index: number;
  }> = [];
  for (let i = 0; i < input.length; ++i) {
    const char = input[i];
    if (openChars.includes(char)) {
      stack.push({
        char,
        index: i,
      });
    } else if (closeChars.includes(char)) {
      if (!stack.length) {
        index = i;
        break;
      } else {
        const { char: lastOpenChar } = stack[stack.length - 1];
        if (char === openAndCloseCharPairs[lastOpenChar]) {
          stack.pop();
        } else {
          ({ index } = stack[0]);
          break;
        }
      }
    }
  }
  if (stack.length) {
    ({ index } = stack[0]);
  }
  if (index === -1) {
    index = input.length;
  }

  while (index > 1 && unreasonableSuffix.includes(input[index - 1])) {
    index--;
  }
  return input.slice(0, index);
};

/**
 * 分割字符串中的超链接和常规字符，输出格式化数据 HyperNode[]
 */
export const getHyperLinkNodes = (str: string): HyperNode[] => {
  const nodes: HyperNode[] = [];
  let cur = 0;
  urlRegExp.lastIndex = 0;
  while (true) {
    const result = urlRegExp.exec(str);
    if (!result) {
      break;
    }

    const { index, 0: content } = result;
    const text = str.slice(cur, index);

    // ① 文字部分
    if (text) {
      nodes.push({
        type: HyperNodeType.text,
        text,
      });
    }

    // ② 链接/markdown/email 部分，email 被渲染为文字，code block / inline code 中的链接也当做文字
    if (content) {
      /** 1. 非正常前缀 */
      const abnormalPrefixMatchResult = /^(@|#|\+|-|%)+/.exec(content);
      if (abnormalPrefixMatchResult) {
        nodes.push({
          type: HyperNodeType.text,
          text: abnormalPrefixMatchResult[0],
        });
        cur = result.index + abnormalPrefixMatchResult[0].length;
        urlRegExp.lastIndex = cur;
      } else {
        /** 2. 合理Url作为前缀 */
        const subUrl = getAReasonableUrl(content);

        if (subUrl) {
          if (emailValidator.validate(subUrl)) {
            nodes.push({
              type: HyperNodeType.text,
              text: subUrl,
            });
          } else {
            nodes.push({
              type: HyperNodeType.link,
              url: subUrl,
            });
          }

          cur = result.index + subUrl.length;
          urlRegExp.lastIndex = cur;
        } else {
          const skipStr = content.slice(0, 1);

          nodes.push({
            type: HyperNodeType.text,
            text: skipStr,
          });

          cur = result.index + skipStr.length;
          urlRegExp.lastIndex = cur;
        }
      }
    }
  }
  if (cur !== str.length) {
    nodes.push({ type: HyperNodeType.text, text: str.slice(cur) });
  }
  return nodes;
};
