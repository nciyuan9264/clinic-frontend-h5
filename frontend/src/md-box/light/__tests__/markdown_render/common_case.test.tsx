import { render, act } from '@testing-library/react';

import { MdBoxForTesting } from '../utils';
import { markdownPlaygroundExample } from '../../../../_internal';

describe('各种不能被纳入其他文件的、较综合的测试用例', () => {
  test('Markdown综合快照测试', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={markdownPlaygroundExample}
          autoFixSyntax={{ autoFixEnding: false }}
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();
  });
});
