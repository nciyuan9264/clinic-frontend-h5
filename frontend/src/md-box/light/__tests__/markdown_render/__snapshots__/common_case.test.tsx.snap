// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`各种不能被纳入其他文件的、较综合的测试用例 Markdown综合快照测试 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="false"
    dir="ltr"
    theme-mode="light"
  >
    <h1
      class="header auto-hide-last-sibling-br"
    >
      My Markdown Document
    </h1>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Introduction
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      Hello, and welcome to my Markdown document! This document will demonstrate some of the features of Markdown, including GFM syntax.
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Headings
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      In Markdown, you can create headings by adding one or more hash symbols (#) before your text. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h1
      class="header auto-hide-last-sibling-br"
    >
      Heading 1
    </h1>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Heading 2
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <h3
      class="header auto-hide-last-sibling-br"
    >
      Heading 3
    </h3>
    <br
      class="container wrapper non-select"
    />
    

    <h4
      class="header auto-hide-last-sibling-br"
    >
      Heading 4
    </h4>
    <br
      class="container wrapper non-select"
    />
    

    <h5
      class="header auto-hide-last-sibling-br"
    >
      Heading 5
    </h5>
    <br
      class="container wrapper non-select"
    />
    

    <h6
      class="header auto-hide-last-sibling-br"
    >
      Heading 6
    </h6>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Paragraphs
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      You can create paragraphs by simply typing your text on separate lines. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      This is a paragraph.
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      This is another paragraph.
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Emphasis
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      You can add emphasis to your text by using asterisks or underscores. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      <em>
        This text will be italicized.
      </em>
      <br
        class="container wrapper non-select"
      />
      <em>
        This text will also be italicized.
      </em>
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      <strong>
        This text will be bold.
      </strong>
      <br
        class="container wrapper non-select"
      />
      <strong>
        This text will also be bold.
      </strong>
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Lists
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      You can create ordered and unordered lists in Markdown. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h3
      class="header auto-hide-last-sibling-br"
    >
      Unordered List
    </h3>
    <br
      class="container wrapper non-select"
    />
    

    <ul
      class="auto-hide-last-sibling-br"
    >
      

      <li>
        Item 1
      </li>
      

      <li>
        Item 2
      </li>
      

      <li>
        Item 3
      </li>
      

    </ul>
    <br
      class="container wrapper non-select"
    />
    

    <h3
      class="header auto-hide-last-sibling-br"
    >
      Ordered List
    </h3>
    <br
      class="container wrapper non-select"
    />
    

    <ol
      class="auto-hide-last-sibling-br"
    >
      

      <li>
        Item 1
      </li>
      

      <li>
        Item 2
      </li>
      

      <li>
        Item 3
      </li>
      

    </ol>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Code Blocks
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      You can create code blocks in Markdown by using backticks or triple backticks. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h3
      class="header auto-hide-last-sibling-br"
    >
      Inline Code
    </h3>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      To use a variable in JavaScript, you can use the 
      <code>
        var
      </code>
       keyword.
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h3
      class="header auto-hide-last-sibling-br"
    >
      Code Block
    </h3>
    <br
      class="container wrapper non-select"
    />
    <div
      class="container code-group-item--normal hide-indicator"
      data-testid="code_block"
    >
      <div
        class="code-area"
        dir="ltr"
      >
        <div
          class="header"
        >
          <div
            class="text"
          >
            <div
              class="icon"
            />
            plaintext
          </div>
          <div
            class="actions"
          >
            <div
              class="item"
            >
              <div
                class="icon hoverable"
                data-testid="code_block_copy"
              >
                <div
                  class="img"
                />
              </div>
            </div>
            <div
              class="item"
            >
              <div
                class="icon hoverable"
                data-testid="code_block_mode_switch"
              >
                <div
                  class="img"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="content content--plaintext"
        >
          <pre
            class="container language-plaintext"
          >
            <code
              class="language-plaintext"
            >
              function addNumbers(a, b) {
  return a + b;
}

            </code>
          </pre>
        </div>
      </div>
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h3
      class="header auto-hide-last-sibling-br"
    >
      Fenced Code Block
    </h3>
    <br
      class="container wrapper non-select"
    />
    <div
      class="container code-group-item--normal hide-indicator"
      data-testid="code_block"
    >
      <div
        class="code-area"
        dir="ltr"
      >
        <div
          class="header"
        >
          <div
            class="text"
          >
            <div
              class="icon"
            />
            javascript
          </div>
          <div
            class="actions"
          >
            <div
              class="item"
            >
              <div
                class="icon hoverable"
                data-testid="code_block_copy"
              >
                <div
                  class="img"
                />
              </div>
            </div>
            <div
              class="item"
            >
              <div
                class="icon hoverable"
                data-testid="code_block_mode_switch"
              >
                <div
                  class="img"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="content content--javascript"
        >
          <pre
            class="container language-javascript"
          >
            <code
              class="language-javascript"
            >
              <span
                class="token keyword"
              >
                function
              </span>
               
              <span
                class="token function"
              >
                addNumbers
              </span>
              <span
                class="token punctuation"
              >
                (
              </span>
              <span
                class="token parameter"
              >
                a
                <span
                  class="token punctuation"
                >
                  ,
                </span>
                 b
              </span>
              <span
                class="token punctuation"
              >
                )
              </span>
               
              <span
                class="token punctuation"
              >
                {
              </span>
              
  
              <span
                class="token keyword"
              >
                return
              </span>
               a 
              <span
                class="token operator"
              >
                +
              </span>
               b
              <span
                class="token punctuation"
              >
                ;
              </span>
              

              <span
                class="token punctuation"
              >
                }
              </span>
              

            </code>
          </pre>
        </div>
      </div>
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Task Lists
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      You can create task lists in Markdown by using brackets. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <ul
      class="tasklist auto-hide-last-sibling-br"
    >
      

      <li>
        <input
          checked=""
          disabled=""
          type="checkbox"
        />
         Write example document
      </li>
      

      <li>
        <input
          disabled=""
          type="checkbox"
        />
         Publish document
      </li>
      

      <li>
        <input
          disabled=""
          type="checkbox"
        />
         Share document with friends
      </li>
      

    </ul>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Tables
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      You can create tables in Markdown using pipes and dashes. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br table-container"
    >
      <table>
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Age
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              John
            </td>
            <td>
              27
            </td>
          </tr>
          <tr>
            <td>
              Jane
            </td>
            <td>
              25
            </td>
          </tr>
          <tr>
            <td>
              Jack
            </td>
            <td>
              30
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Links
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      You can create links in Markdown using brackets and parentheses. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      <a
        class="link"
        href="https://www.google.com"
        target="_blank"
      >
        Google
      </a>
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Images
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      You can add images to your Markdown document using exclamation marks, brackets, and parentheses. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      <div
        class="container"
      >
        <div
          class="image-wrapper"
          data-testid="mdbox_image"
          style="width: 100%; height: 0px; max-width: 400px;"
        >
          <span
            style="display: inline-block;"
          >
            <img
              alt="image"
              class="image"
              decoding="async"
              height="256"
              imagex-type="react"
              imagex-version="0.3.7"
              loading="lazy"
              src="https://pic1.zhimg.com/v2-b444070848d54baf536222b22a51fba4_b.jpg"
              style="border-radius: 12px; overflow: hidden; filter: blur(20px); background-size: cover; background-image: url(data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTAwJSIgaGVpZ2h0PSIxMDAlIiB2aWV3Qm94PSIwLDAsMjAlLDEwMCUiIHZlcnNpb249IjEuMSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB4bWxuczp4bGluaz0iaHR0cDovL3d3dy53My5vcmcvMTk5OS94bGluayI+CiAgPGRlZnM+CiAgICA8bGluZWFyR3JhZGllbnQgaWQ9ImciPgogICAgICA8c3RvcCBzdG9wLWNvbG9yPSIjRjJGM0Y1IiBvZmZzZXQ9IjI1JSIgLz4KICAgICAgPHN0b3Agc3RvcC1jb2xvcj0iI0U1RTZFQiIgb2Zmc2V0PSIzNyUiIC8+CiAgICAgIDxzdG9wIHN0b3AtY29sb3I9IiNGMkYzRjUiIG9mZnNldD0iNjMlIiAvPgogICAgPC9saW5lYXJHcmFkaWVudD4KICA8L2RlZnM+CiAgPHJlY3QgaWQ9InIiIHdpZHRoPSI0MDAlIiBoZWlnaHQ9IjEwMCUiIGZpbGw9InVybCgjZykiIC8+CiAgPGFuaW1hdGUgeGxpbms6aHJlZj0iI3IiIGF0dHJpYnV0ZU5hbWU9IngiIGZyb209Ii0zMDAlIiB0bz0iMCUiIGR1cj0iMS41cyIgcmVwZWF0Q291bnQ9ImluZGVmaW5pdGUiICAvPgo8L3N2Zz4=); background-position: center;"
              width="400"
            />
          </span>
        </div>
      </div>
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Blockquotes
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      You can create blockquotes in Markdown using the greater than symbol. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <blockquote
      class="auto-hide-last-sibling-br"
    >
      

      <div
        class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
      >
        This is a blockquote.
      </div>
      <br
        class="container wrapper non-select"
      />
      

    </blockquote>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Horizontal Rule
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      You can add a horizontal rule in Markdown using three or more hyphens, asterisks, or underscores. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <hr
      class="auto-hide-last-sibling-br"
    />
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Emojis
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      You can add emojis to your Markdown document using colons and the name of the emoji. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      :smiley:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Strikethrough
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      You can add strikethrough text to your Markdown document using two tildes. For example:
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      <del>
        This text is strikethrough.
      </del>
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <h2
      class="header auto-hide-last-sibling-br"
    >
      Conclusion
    </h2>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      That's it for my Markdown document! I hope you found it helpful. Thanks for reading!
    </div>
    <br
      class="container wrapper non-select"
    />
  </div>
</DocumentFragment>
`;
